## Part 1: Preprocessing

In [3]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import numpy as np
from tensorflow.keras.models import Model
from tensorflow.keras import layers

#  Import and read the attrition data
attrition_df = pd.read_csv('https://static.bc-edx.com/ai/ail-v-1-0/m19/lms/datasets/attrition.csv')
attrition_df.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   41       Yes      Travel_Rarely                   Sales                 1   
1   49        No  Travel_Frequently  Research & Development                 8   
2   37       Yes      Travel_Rarely  Research & Development                 2   
3   33        No  Travel_Frequently  Research & Development                 3   
4   27        No      Travel_Rarely  Research & Development                 2   

   Education EducationField  EnvironmentSatisfaction  HourlyRate  \
0          2  Life Sciences                        2          94   
1          1  Life Sciences                        3          61   
2          2          Other                        4          92   
3          4  Life Sciences                        4          56   
4          1        Medical                        1          40   

   JobInvolvement  ...  PerformanceRating RelationshipSatisfaction  \
0               3  ...                  3                        1   
1               2  ...                  4                        4   
2               2  ...                  3                        2   
3               3  ...                  3                        3   
4               3  ...                  3                        4   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  \
0                 0                 8                      0               1   
1                 1                10                      3               3   
2                 0                 7                      3               3   
3                 0                 8                      3               3   
4                 1                 6                      3               3   

   YearsAtCompany  YearsInCurrentRole  YearsSinceLastPromotion  \
0               6                   4                        0   
1              10                   7                        1   
2               0                   0                        0   
3               8                   7                        3   
4               2                   2                        2   

   YearsWithCurrManager  
0                     5  
1                     7  
2                     0  
3                     0  
4                     2  

[5 rows x 27 columns]

In [30]:
#evaluate the data for the attrition column based on question involving accuracy (futher down after running the models)
attrition_df["Attrition"].value_counts()

No     1233
Yes     237
Name: Attrition, dtype: int64

In [4]:
# Determine the number of unique values in each column.
attrition_df.nunique()

Age                         43
Attrition                    2
BusinessTravel               3
Department                   3
DistanceFromHome            29
Education                    5
EducationField               6
EnvironmentSatisfaction      4
HourlyRate                  71
JobInvolvement               4
JobLevel                     5
JobRole                      9
JobSatisfaction              4
MaritalStatus                3
NumCompaniesWorked          10
OverTime                     2
PercentSalaryHike           15
PerformanceRating            2
RelationshipSatisfaction     4
StockOptionLevel             4
TotalWorkingYears           40
TrainingTimesLastYear        7
WorkLifeBalance              4
YearsAtCompany              37
YearsInCurrentRole          19
YearsSinceLastPromotion     16
YearsWithCurrManager        18
dtype: int64

In [4]:
# I used the approach we learned in our activities which involved doing our preprocessing prior to splitting the data.

In [5]:
#Preprocess Y
from sklearn.preprocessing import OneHotEncoder, LabelEncoder

In [6]:
#encode y - Attrition and Department columns
enc_attrition = OneHotEncoder(sparse_output=False)
attrition_encoded = enc_attrition.fit_transform(attrition_df[["Attrition"]])
attrition_columns = enc_attrition.get_feature_names_out(["Attrition"])
df_attrition = pd.DataFrame(attrition_encoded, columns= attrition_columns)


In [7]:
enc_department = OneHotEncoder(sparse_output=False)
department_encoded = enc_department.fit_transform(attrition_df[["Department"]])
department_columns = enc_department.get_feature_names_out(["Department"])
df_department = pd.DataFrame(department_encoded, columns= department_columns)

In [8]:
#concat to original data frame - based on what we'd done previously I added the new columns to our dataframe and removed the originals
#Further work identified I didn't need the concated dataframe so it's not used other than creation of my y_df that is already encoded.
encoded_df = pd.concat([attrition_df,df_department,df_attrition], axis =1)

In [9]:
#drop original "Attrition and Department" columns
encoded_df = encoded_df.drop(["Attrition","Department"], axis=1)

In [10]:
# Create y_df with the Attrition and Department columns

y_attrition= encoded_df[["Attrition_No","Attrition_Yes"]]
y_department = encoded_df[["Department_Sales","Department_Human Resources", "Department_Research & Development"]]
#convert the data to an array
y_attrition = y_attrition.to_numpy()
y_department = y_department.to_numpy()


In [13]:
# Create a list of at least 10 column names to use as X data
columns = attrition_df[['Age', 'DistanceFromHome', 'Education', 'EnvironmentSatisfaction', 'HourlyRate', 'JobInvolvement', 'JobLevel', 'JobSatisfaction', 'NumCompaniesWorked', 'PercentSalaryHike', 'PerformanceRating', 'RelationshipSatisfaction', 'StockOptionLevel', 'TotalWorkingYears', 'TrainingTimesLastYear', 'WorkLifeBalance', 'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion', 'YearsWithCurrManager']]


# Create X_df using your selected columns
X_df = pd.DataFrame(columns)

# Show the data types for X_df
print(X_df.dtypes)
X_df.head()


Age                         int64
DistanceFromHome            int64
Education                   int64
EnvironmentSatisfaction     int64
HourlyRate                  int64
JobInvolvement              int64
JobLevel                    int64
JobSatisfaction             int64
NumCompaniesWorked          int64
PercentSalaryHike           int64
PerformanceRating           int64
RelationshipSatisfaction    int64
StockOptionLevel            int64
TotalWorkingYears           int64
TrainingTimesLastYear       int64
WorkLifeBalance             int64
YearsAtCompany              int64
YearsInCurrentRole          int64
YearsSinceLastPromotion     int64
YearsWithCurrManager        int64
dtype: object


Age  DistanceFromHome  Education  EnvironmentSatisfaction  HourlyRate  \
0   41                 1          2                        2          94   
1   49                 8          1                        3          61   
2   37                 2          2                        4          92   
3   33                 3          4                        4          56   
4   27                 2          1                        1          40   

   JobInvolvement  JobLevel  JobSatisfaction  NumCompaniesWorked  \
0               3         2                4                   8   
1               2         2                2                   1   
2               2         1                3                   6   
3               3         1                3                   1   
4               3         1                2                   9   

   PercentSalaryHike  PerformanceRating  RelationshipSatisfaction  \
0                 11                  3                         1   
1                 23                  4                         4   
2                 15                  3                         2   
3                 11                  3                         3   
4                 12                  3                         4   

   StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
0                 0                  8                      0   
1                 1                 10                      3   
2                 0                  7                      3   
3                 0                  8                      3   
4                 1                  6                      3   

   WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  \
0                1               6                   4   
1                3              10                   7   
2                3               0                   0   
3                3               8                   7   
4                3               2                   2   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     5  
1                        1                     7  
2                        0                     0  
3                        3                     0  
4                        2                     2

In [14]:
# Split the data into training and testing sets
from sklearn.model_selection import train_test_split
X_train,X_test,y_atttrition_train,y_attrition_test, y_department_train, y_department_test = train_test_split(X_df,y_attrition,y_department,random_state=78)


In [15]:
# Convert your X data to numeric data types however you see fit
#NOT NECESSARY - X DATA ALREADY IN NUMERIC FORM
# Add new code cells as necessary


In [16]:
# Create a StandardScaler
scaler = StandardScaler()

# Fit the StandardScaler to the training data
scaler.fit(X_train)

# Scale the training and testing data
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)


In [17]:
#ensure data is in array format
print(X_train_scaled)

[[-0.40942386  0.87475435  0.10528682 ... -0.590081   -0.67430453
  -0.58065358]
 [-1.29889668 -0.50720476 -0.88638904 ... -0.31430947 -0.34506944
   0.26463076]
 [-0.07587155  0.49785641  2.08863855 ... -0.590081   -0.01583436
  -0.58065358]
 ...
 [-0.18705565 -0.63283741 -0.88638904 ... -1.14162405 -0.67430453
  -1.14417648]
 [-1.41008078  1.87981553  0.10528682 ... -0.590081   -0.01583436
  -0.86241503]
 [-0.63179206 -0.00467417 -0.88638904 ... -0.590081   -0.34506944
  -0.58065358]]


## Create, Compile, and Train the Model

In [18]:
from tensorflow.keras.optimizers import Adam

In [19]:
# Find the number of columns in the X training data
#Because I created a list of more than 10 features my output for numbers will be different from the example
num_cols = X_train.shape[1]
print(num_cols)

# Create the input layer
input_layer = layers.Input(shape = (X_train.shape[1],), name="input_features")

# Create at least two shared layers
shared_layer1=layers.Dense(64, activation="relu")(input_layer)
shared_layer2=layers.Dense(128, activation = "relu")(shared_layer1)

20


In [20]:
# Create a branch for Department
# with a hidden layer and an output layer
# Create the hidden layer
department_hidden = layers.Dense(units=32, activation='relu', name="department_hidden")(shared_layer2)

# Create the output layer
department_output = layers.Dense(units=3, activation='softmax', name="department_output")(department_hidden)



In [21]:
# Create a branch for Attrition
# with a hidden layer and an output layer

# Create the hidden layer
attrition_hidden = layers.Dense(units=32, activation='relu', name="attrition_hidden")(shared_layer2)

# Create the output layer
attrition_output = layers.Dense(units=2, activation='softmax', name="attrition_output")(attrition_hidden)


In [22]:
from tensorflow.keras.optimizers import Adam
model = Model(inputs=input_layer, outputs=[attrition_output, department_output])
# Compile the model
model.compile(optimizer=Adam(),
              loss={"attrition_output": "binary_crossentropy", "department_output": "categorical_crossentropy"},
              metrics={"attrition_output": "accuracy", "department_output": "accuracy"})
# Summarize the model
model.summary()

Model: "model"
__________________________________________________________________________________________________
 Layer (type)                Output Shape                 Param #   Connected to                  
 input_features (InputLayer  [(None, 20)]                 0         []                            
 )                                                                                                
                                                                                                  
 dense (Dense)               (None, 64)                   1344      ['input_features[0][0]']      
                                                                                                  
 dense_1 (Dense)             (None, 128)                  8320      ['dense[0][0]']               
                                                                                                  
 attrition_hidden (Dense)    (None, 32)                   4128      ['dense_1[0][0]']         

In [23]:
# Train the model
model.fit(X_df,
          {"attrition_output":y_attrition, "department_output":y_department},
          epochs=100,
          batch_size = 32,
          validation_split = 0.2)


Epoch 1/100
37/37 [==============================] - 3s 17ms/step - loss: 1.4722 - attrition_output_loss: 0.5451 - department_output_loss: 0.9271 - attrition_output_accuracy: 0.8367 - department_output_accuracy: 0.5842 - val_loss: 1.4257 - val_attrition_output_loss: 0.4202 - val_department_output_loss: 1.0055 - val_attrition_output_accuracy: 0.8707 - val_department_output_accuracy: 0.6054
Epoch 2/100
37/37 [==============================] - 0s 6ms/step - loss: 1.2327 - attrition_output_loss: 0.4324 - department_output_loss: 0.8003 - attrition_output_accuracy: 0.8376 - department_output_accuracy: 0.6420 - val_loss: 1.4433 - val_attrition_output_loss: 0.4069 - val_department_output_loss: 1.0365 - val_attrition_output_accuracy: 0.8741 - val_department_output_accuracy: 0.3878
Epoch 3/100
37/37 [==============================] - 0s 5ms/step - loss: 1.1919 - attrition_output_loss: 0.4197 - department_output_loss: 0.7722 - attrition_output_accuracy: 0.8359 - department_output_accuracy: 0.6369

In [24]:
# Evaluate the model with the testing data
test_results = model.evaluate(X_test, {"attrition_output": y_attrition_test, "department_output": y_department_test})
test_results


12/12 [==============================] - 0s 3ms/step - loss: 0.7429 - attrition_output_loss: 0.1908 - department_output_loss: 0.5521 - attrition_output_accuracy: 0.9348 - department_output_accuracy: 0.7935


[0.7429246306419373,
 0.1908065378665924,
 0.552118182182312,
 0.9347826242446899,
 0.79347825050354]

In [27]:
# Print the accuracy for both department and attrition
print(f"Attrition Accuracy: {test_results[3]}")
print(f"Department Accuracy: {test_results[4]}")

Attrition Accuracy: 0.9347826242446899
Department Accuracy: 0.79347825050354


# Summary

In the provided space below, briefly answer the following questions.

1. Is accuracy the best metric to use on this data? Why or why not?

2. What activation functions did you choose for your output layers, and why?

3. Can you name a few ways that this model might be improved?

YOUR ANSWERS HERE

1.Accuracy may not be the best choice because our data may be imbalanced.  When I review the data for this column, we have many more no answers than yes (No-1233 Yes -237). We would probably want to rerun the model after balancing the data.
2.I used relu for the binary value of the attrition out-put.  I could have used sigmoid, but thought I'd use relu instead.  For the department output, because it was a multi-class I used the softmax activation function.
3.Yes - we could address the imbalance of attrition data, through oversampling; we could hypertune the models and add more neurons and layers. We could feature engineer if there are columns that make sense for that and also use K-fold and cross validation to further refine the data.